controllers/provider_controller.go

```diff
@@ -44,61 +45,107 @@ import (
+	r.RecordSuspend(ctx, provider, provider.Spec.Suspend)
+	// return early if the object is suspended
+	if provider.Spec.Suspend {
+		log.Info("Reconciliation is suspended for this object")
+		return ctrl.Result{}, nil
 	}
 
-	// validate provider spec and credentials
-	if err := r.validate(ctx, provider); err != nil {
-		meta.SetResourceCondition(&provider, meta.ReadyCondition, metav1.ConditionFalse, meta.ReconciliationFailedReason, err.Error())
-		if err := r.patchStatus(ctx, req, provider.Status); err != nil {
-			return ctrl.Result{Requeue: true}, err
-		}
-		return ctrl.Result{Requeue: true}, err
+	patchHelper, err := patch.NewHelper(provider, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
 	}
 
-	if !apimeta.IsStatusConditionTrue(provider.Status.Conditions, meta.ReadyCondition) {
-		meta.SetResourceCondition(&provider, meta.ReadyCondition, metav1.ConditionTrue, v1beta1.InitializedReason, v1beta1.InitializedReason)
-		if err := r.patchStatus(ctx, req, provider.Status); err != nil {
-			return ctrl.Result{Requeue: true}, err
+	defer func() {
+		patchOpts := []patch.Option{
+			patch.WithOwnedConditions{
+				Conditions: []string{
+					meta.ReadyCondition,
+					meta.ReconcilingCondition,
+					meta.StalledCondition,
+				},
+			},
 		}
-		log.Info("Provider initialised")
-	}
 
-	r.recordReadiness(ctx, provider)
+		if retErr == nil && (result.IsZero() || !result.Requeue) {
+			conditions.Delete(provider, meta.ReconcilingCondition)
 
-	return ctrl.Result{}, nil
+			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
+
+			readyCondition := conditions.Get(provider, meta.ReadyCondition)
+			switch readyCondition.Status {
+			case metav1.ConditionFalse:
+				// As we are no longer reconciling and the end-state is not ready, the reconciliation has stalled
+				conditions.MarkStalled(provider, readyCondition.Reason, readyCondition.Message)
+			case metav1.ConditionTrue:
+				// As we are no longer reconciling and the end-state is ready, the reconciliation is no longer stalled
+				conditions.Delete(provider, meta.StalledCondition)
+			}
+		}
+
+		if err := patchHelper.Patch(ctx, provider, patchOpts...); err != nil {
+			retErr = kerrors.NewAggregate([]error{retErr, err})
+		}
+
+		r.Metrics.RecordReadiness(ctx, provider)
+		r.Metrics.RecordDuration(ctx, provider, start)
+
+	}()
+
+	return r.reconcile(ctx, provider)
 }
 
-func (r *ProviderReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&v1beta1.Provider{}).
-		WithEventFilter(predicate.Or(predicate.GenerationChangedPredicate{}, predicates.ReconcileRequestedPredicate{})).
-		Complete(r)
+func (r *ProviderReconciler) reconcile(ctx context.Context, obj *v1beta1.Provider) (ctrl.Result, error) {
+	// Mark the resource as under reconciliation
+	conditions.MarkReconciling(obj, meta.ProgressingReason, "")
+
+	// validate provider spec and credentials
+	if err := r.validate(ctx, obj); err != nil {
```

Comment:
As this validates both the spec, and references itself, I think it would be wise to filter the error. As invalid spec errors do not require a requeue, given we do not expect to recover without a generation change of the object.

Reply:
I think that in a case where the secret(`provider.Spec.SecretRef`) is missing, the secret could be created and this could cause a recover without a generation change of the object.

Reply:
You are absolutely right!